README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
zFactor.DL
==========

The goal of zFactor.DL is to create a correlation using neural networks.

Dataset
-------

These observations have been taken from a digital scan of the 1942 Standing-Katz chart for hydrocarbon gases.

### Original Standing-Katz dataset

The original data was gracefully provided by the authors of the paper \[Kamyab, M., et al., Using artificial neural networks to estimate the z-factor for natural hydrocarbon gases, J. Pet. Sci.Eng. (2010), <doi:10.1016/j.petrol.2010.07.006>\] (<http://www.sciencedirect.com/science/article/pii/S0920410510001427?via%3Dihub>).

The original dataset, for Ppr from 0 to 15, is an Excel file named `SK_data.xls` with Ppr, Tpr and z in columnar format with 2853 rows and 60 columns. It is under the folder \[./inst/extdata\] (<https://github.com/f0nzie/zFactor.DL/tree/master/inst/extdata>) in this repository. They are three columns or set per Tpr curve.

    row    Ppr.0 Tpr.0 z.0    Ppr.1 Tpr.1 z.1  ...    Ppr.19 Tpr.19 z.19
     1
     . 
     .
     .
    2853

### Tidy Standing-Katz dataset

The tidy dataset contains 57964 observations of compressibility factors at different Pseudo-reduced Temperatures and Pseudo-reduced pressures. The original data was converted to tidy format using `R`. Refer to notebook [01-read\_raw\_data.Rmd](https://github.com/f0nzie/zFactor.DL/blob/master/notebooks/01-read_raw_data.Rmd) for step by step process.

The original file has been converted to a tidy format file named `tidy_SK.csv` with 57,060 observations and 4 variables with the column names Tpr, Ppr, z and set number. This file is under the same folder \[./inst/extdata\] (<https://github.com/f0nzie/zFactor.DL/tree/master/inst/extdata>).

The tidy dataset format looks like this, 57,060 observations (rows) and 4 variables (columns):

    row     Tpr   Ppr   z   set
       1       
       2
       . 
       .
       .
    57060

``` r
data("sk_tidy")
```

``` r
dim(sk_tidy)
#> [1] 57964     5
```

``` r
names(sk_tidy)
#> [1] "Tpr"   "Ppr"   "z"     "set"   "range"
```

``` r
min(sk_tidy$Tpr)
#> [1] 1.05
max(sk_tidy$Tpr)
#> [1] 3

min(sk_tidy$Ppr)
#> [1] 0
max(sk_tidy$Ppr)
#> [1] 30

min(sk_tidy$z)
#> [1] 0.251754
max(sk_tidy$z)
#> [1] 2.66
```

``` r
summary(sk_tidy)
#>       Tpr             Ppr               z               set        
#>  Min.   :1.050   Min.   : 0.000   Min.   :0.2518   Min.   : 0.000  
#>  1st Qu.:1.300   1st Qu.: 2.558   1st Qu.:0.8477   1st Qu.: 4.000  
#>  Median :1.600   Median : 5.680   Median :0.9921   Median : 9.000  
#>  Mean   :1.743   Mean   : 7.008   Mean   :1.0335   Mean   : 9.406  
#>  3rd Qu.:2.200   3rd Qu.:11.853   3rd Qu.:1.2621   3rd Qu.:14.000  
#>  Max.   :3.000   Max.   :30.000   Max.   :2.6600   Max.   :19.000  
#>     range          
#>  Length:57964      
#>  Class :character  
#>  Mode  :character  
#>                    
#>                    
#> 
```

``` r
data("sk_tidy_lp")
summary(sk_tidy_lp)
#>       Tpr             Ppr               z              set           
#>  Min.   :1.050   Min.   : 0.000   Min.   :0.2518   Length:57060      
#>  1st Qu.:1.288   1st Qu.: 2.525   1st Qu.:0.8439   Class :character  
#>  Median :1.550   Median : 5.577   Median :0.9882   Mode  :character  
#>  Mean   :1.738   Mean   : 6.754   Mean   :1.0214                     
#>  3rd Qu.:2.050   3rd Qu.:11.647   3rd Qu.:1.2509                     
#>  Max.   :3.000   Max.   :15.001   Max.   :1.7536
```

``` r
data("sk_tidy_hp")
summary(sk_tidy_hp)
#>       Tpr            Ppr             z             set           
#>  Min.   :1.40   Min.   :16.0   Min.   :1.363   Length:904        
#>  1st Qu.:1.75   1st Qu.:19.5   1st Qu.:1.585   Class :character  
#>  Median :2.10   Median :23.0   Median :1.749   Mode  :character  
#>  Mean   :2.10   Mean   :23.0   Mean   :1.792                     
#>  3rd Qu.:2.45   3rd Qu.:26.5   3rd Qu.:1.945                     
#>  Max.   :2.80   Max.   :30.0   Max.   :2.660
```


NEWS.md


# zFactor.DL 0.1.0
* total of `57964` rows and and `5` columns, combination of low and high pressure values.
* variable `set` converted to numeric in `sk_tidy.rda`.
* replace `sk_tidy.rda` with a combined table of low and high pressure.
* make `set` column in both tables numeric.
* add column `range` to both tables indicating if `lo` or `hi`.
* create dataset `sk_tidy_lp.rda` for low pressure
* create dataset `sk_tidy_hp.rda` for high pressure
* read high pressure (Ppr) data points for z. use same HP notebook for low pressure. now we have two .Rda datasets in data.
* add extra notebook to read high pressure (Ppr) data points for z.
* added github logo to website
* Added pkgdown website to repo.
* sending website in `docs` folder to `gh-pages`.
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(zFactor.DL)

test_check("zFactor.DL")


tests/testthat/test_tidy_vs_raw_data.R
library(testthat)


test_that("one point in A:C,2854 yields true", {
    # point read from raw data table
      ac2854 <- c(15.0008,	1.05,	1.75359)
    # logical conditions
      mask <- sk_tidy$Ppr == ac2854[1] & sk_tidy$Tpr == ac2854[2]
    # assert the lookup value
    expect_true(sk_tidy[which(mask), "z"] == ac2854[3])

})


test_that("one point in A:C,2854 yields true", {
    # point read from raw data table
    bfbh2854 <- c(15.0008,	3.0,	1.33488)
    # logical conditions
    mask <- sk_tidy$Ppr == bfbh2854[1] & sk_tidy$Tpr == bfbh2854[2]
    # assert the lookup value
      expect_true(sk_tidy[which(mask), "z"] == bfbh2854[3])

})

test_that("multiple z points in raw dataset match z in tidy dataset", {
    zList <- list(
        c(0.00402266,	1.15,	1.00046),  # (3, G:I)
        c(0.00402266,	3,   	1.00046),  # (3, BF:BH)
        c(0.00402266,	1.6,	1.00046),  # (3, AE:AG)
        c(4.81921,   	1.05,	0.639045), # (1250, A:C)
        c(4.81921,	    3,	    1.04625),  # (1250, BF:BH)
        c(4.81921,	    1.6,	0.839743)  # (1250, AE:AG)
    )

    for (arow in zList) {
        mask <- sk_tidy$Ppr == arow[1] & sk_tidy$Tpr == arow[2]
        # cat(sprintf("%f %f \n", SK_tidy[which(mask), "z"], arow[3]))
        expect_true(sk_tidy[which(mask), "z"] == arow[3])
    }
})


f <- function(x) {
    # function to compare a row of the dataset z values
    # dataset vs raw data
    mask <- sk_tidy$Ppr == x[1] & sk_tidy$Tpr == x[2]
    # cat(sprintf("%f %f \n", SK_tidy[which(mask), "z"], x[3]))
    expect_true(sk_tidy[which(mask), "z"] == x[3])
}

test_that("chunks of raw data points match tidy dataset", {
    raw_chunk <- "
    0.677651	1.05	0.758361
    0.678217	1.05	0.758079
    0.685469	1.05	0.754464
    0.685493	1.05	0.754452
    0.685895	1.05	0.754251
    0.686456	1.05	0.753972
    0.686649	1.05	0.753875
    0.695518	1.05	0.749454
    0.702943	1.05	0.746121
    0.703360    1.05    0.745934
    0.711643    1.05    0.742216
    "
    raw_points <- read.table(text = raw_chunk, header = FALSE, sep = "")
    apply(raw_points, 1, f)
})

test_that("bigger chunks of raw data points match the tidy dataset", {

raw_chunk <- "
13.0085	1.05	1.54907
13.0091	1.05	1.54913
13.0093	1.05	1.54915
13.0185	1.05	1.55007
13.0262	1.05	1.55084
13.0269	1.05	1.55091
13.027	1.05	1.55092
13.0347	1.05	1.55169
13.0353	1.05	1.55175
13.0356	1.05	1.55178
13.0358	1.05	1.55179
13.0432	1.05	1.55254
13.0441	1.05	1.55263
13.0443	1.05	1.55265
13.0523	1.05	1.55345
13.0526	1.05	1.55348
13.0601	1.05	1.55423
13.0689	1.05	1.55525
13.0695	1.05	1.55532
13.0783	1.05	1.55636
13.0866	1.05	1.55732
13.0951	1.05	1.55831
13.0962	1.05	1.55844
13.1045	1.05	1.55942
13.1046	1.05	1.55942

13.0093	1.1	1.51636
13.0185	1.1	1.51728
13.0262	1.1	1.51804
13.0269	1.1	1.51811
13.027	1.1	1.51812
13.0347	1.1	1.51889
13.0353	1.1	1.51895
13.0356	1.1	1.51898
13.0358	1.1	1.519
13.0432	1.1	1.51974
13.0441	1.1	1.51983
13.0443	1.1	1.51985
13.0523	1.1	1.52066
13.0526	1.1	1.52068

13.2512	1.25	1.45856
13.2595	1.25	1.4594
13.2596	1.25	1.45941
13.26	1.25	1.45945
13.2678	1.25	1.46022
13.2685	1.25	1.4603
13.2686	1.25	1.4603
13.2762	1.25	1.46107
13.2766	1.25	1.4611
13.2847	1.25	1.46181
13.2932	1.25	1.46255
13.3014	1.25	1.46328
13.3103	1.25	1.46405
13.3196	1.25	1.46487
13.3198	1.25	1.46488
13.3201	1.25	1.46491

2.46245	1.1	0.391235
2.46267	1.1	0.391252
2.4655	1.1	0.391464
2.46583	1.1	0.391489
2.47139	1.1	0.391907
2.47336	1.1	0.392054
2.47476	1.1	0.392159
2.48075	1.1	0.392609
2.49754	1.1	0.393869
2.49775	1.1	0.393884
2.4996	1.1	0.394023
2.50535	1.1	0.394454

4.36919	2.6	1.01398
4.37942	2.6	1.01409
4.3944	2.6	1.01426
4.39449	2.6	1.01426
4.39527	2.6	1.01428
4.39571	2.6	1.01429
4.39587	2.6	1.01429
4.40549	2.6	1.01448
4.41218	2.6	1.01462
4.41358	2.6	1.01464
4.414	2.6	1.01465
4.42241	2.6	1.01482
4.43091	2.6	1.01499
4.43738	2.6	1.01512
4.43744	2.6	1.01512

3.2578	1.05	0.437913
3.25913	1.05	0.438068
3.26437	1.05	0.438767
3.27306	1.05	0.439928
3.28436	1.05	0.441435
3.2848	1.05	0.441494
3.29061	1.05	0.442269
3.29319	1.05	0.442613
3.30006	1.05	0.44353

3.25913	1.1	0.469634
3.26437	1.1	0.470233
3.27306	1.1	0.471228
3.28436	1.1	0.472519
3.2848	1.1	0.47257
3.29061	1.1	0.473249
3.29319	1.1	0.473549
3.30006	1.1	0.474352
3.30775	1.1	0.475249
3.30801	1.1	0.475279

3.27306	1.15	0.513074
3.28436	1.15	0.514063
3.2848	1.15	0.514102
3.29061	1.15	0.514611
3.29319	1.15	0.514836
3.30006	1.15	0.515426
3.30775	1.15	0.516085
3.30801	1.15	0.516107
3.30813	1.15	0.516117
3.30936	1.15	0.516223
3.31084	1.15	0.51635
3.32703	1.15	0.517739
3.33353	1.15	0.518296
3.33414	1.15	0.518348
3.33502	1.15	0.518424
3.33652	1.15	0.518553

3.38824	1.25	0.593541
3.394	1.25	0.593767
3.39426	1.25	0.593777
3.39438	1.25	0.593782
3.39561	1.25	0.59383
3.40486	1.25	0.594192
3.41328	1.25	0.594575
3.41427	1.25	0.59462
3.41978	1.25	0.59487
3.4204	1.25	0.594898
3.42128	1.25	0.594939

"
raw_points <- read.table(text = raw_chunk, header = FALSE, sep = "")
apply(raw_points, 1, f)

})


R/hello.R

hello <- function() {
  print("Hello, world!")
}

#' Compressibilty factor dataset
#'
#' A dataset containing the z factor from SK chart for low and high
#' Pseudo-reduced pressure
#'
#' @format A data frame with 57060 rows and 4 variables:
#' \describe{
#'   \item{Tpr}{pseudo reduce temperature}
#'   \item{Ppr}{pseudo reduced pressure}
#'   \item{z}{compressibility factor}
#'   \item{set}{a set of Tpr curve}
#'   \item{range}{if curve is low or high pressure}
#'   ...
#' }
"sk_tidy"


#' untidy dataset
#'
#' The raw dataset read into R with first three column names changed
#'
#' @format A data frame with 2853 rows and 60 variables:
"untidy"


#' Compressibilty factor for Low Pressure data points
#'
#' A dataset containing the z factor from SK chart fow low Ppr
#'
#' @format A data frame with 57060 rows and 4 variables:
#' \describe{
#'   \item{Tpr}{pseudo reduce temperature}
#'   \item{Ppr}{pseudo reduced pressure}
#'   \item{z}{compressibility factor}
#'   \item{set}{a set of Tpr curve}
#'   ...
#' }
"sk_tidy_lp"


#' Compressibilty factor for High Pressure data points
#'
#' A dataset containing the z factor from SK chart fow high Ppr
#'
#' @format A data frame with 904 rows and 4 variables:
#' \describe{
#'   \item{Tpr}{pseudo reduce temperature}
#'   \item{Ppr}{pseudo reduced pressure}
#'   \item{z}{compressibility factor}
#'   \item{set}{a set of Tpr curve}
#'   ...
#' }
"sk_tidy_hp"
